gateway/src/node/client_handling/websocket/connection_handler/ecash/mod.rs

```diff
@@ -0,0 +1,481 @@
+        api_clients: &[CoconutApiClient],
+        credential: &CredentialSpendingRequest,
+    ) -> Result<(), RequestHandlingError> {
+        let serial_number = credential.data.payment.serial_number_bs58();
+        let proposal_id = match credential.data.typ {
+            CredentialType::Voucher => unimplemented!(),
+            CredentialType::TicketBook => {
+                let res = self
+                    .nyxd_client
+                    .write()
+                    .await
+                    .prepare_credential(serial_number, self.address.to_string(), None)
+                    .await?;
+                let proposal_id = find_attribute(&res.logs, "wasm", BANDWIDTH_PROPOSAL_ID)
+                    .ok_or(RequestHandlingError::ProposalIdError {
+                        reason: String::from("proposal id not found"),
+                    })?
+                    .value
+                    .parse::<u64>()
+                    .map_err(|_| RequestHandlingError::ProposalIdError {
+                        reason: String::from("proposal id could not be parsed to u64"),
+                    })?;
+
+                let proposal = self
+                    .nyxd_client
+                    .read()
+                    .await
+                    .query_proposal(proposal_id)
+                    .await?;
+                if !credential.matches_serial_number(&proposal.description)? {
+                    return Err(RequestHandlingError::ProposalIdError {
+                        reason: String::from("proposal has different serial number"),
+                    });
+                }
+                Some(proposal_id)
+            }
+            CredentialType::FreePass => None,
+        };
+
+        let req = nym_api_requests::coconut::VerifyEcashCredentialBody::new(
+            credential.data.clone(),
+            self.address.clone(),
+            proposal_id,
+        );
+        for client in api_clients {
+            let ret = client.api_client.verify_online_credential(&req).await;
+            let client_url = client.api_client.nym_api.current_url();
+            match ret {
+                Ok(VerifyEcashCredentialResponse::Accepted) => {
+                    debug!("Validator at {client_url} accepted the credential");
+                }
+                Ok(response) => {
+                    debug!("Validator at {client_url} didn't accept the credential. Reason : {response}");
+                }
+                Err(err) => {
+                    warn!("Validator at {client_url} could not be reached. There might be a problem with the coconut endpoint: {err}");
+                }
+            }
+        }
+
+        if let Some(proposal_id) = proposal_id {
+            if self
+                .nyxd_client
+                .read()
+                .await
+                .query_proposal(proposal_id)
+                .await?
+                .status
+                == nym_validator_client::nyxd::cw3::Status::Rejected
```

Comment:
there can be more statuses we want to check for. for example if the proposal is still pending because nym-apis are having issues, we shouldn't blindly accept it. imo do a match on all statuses instead

Reply:
TO FIX